From f1d78c489a39971b5aac5d2fc8a39bfa925c3c5d Mon Sep 17 00:00:00 2001
From: Dandan Bi <dandan.bi@intel.com>
Date: Mon, 6 May 2019 16:43:34 +0800
Subject: [PATCH 04/17] MdeModulePkg/DisplayEngine: Zero memory before free
 (CVE-2019-14558)

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=1611

Cc: Liming Gao <liming.gao@intel.com>
Cc: Eric Dong <eric.dong@intel.com>
Cc: Jian J Wang <jian.j.wang@intel.com>
Signed-off-by: Dandan Bi <dandan.bi@intel.com>
Reviewed-by: Eric Dong <eric.dong@intel.com>
Reviewed-by: Jian J Wang <jian.j.wang@intel.com>
[dannf: dropped copyright year change due to conflict]

Origin: https://github.com/tianocore/edk2/commit/f1d78c489a39971b5aac5d2fc8a39bfa925c3c5d
Bug: https://bugzilla.tianocore.org/show_bug.cgi?id=1611
Last-Update: 2020-04-03

Index: edk2-0~20180205.c0d9813c/MdeModulePkg/Universal/DisplayEngineDxe/ProcessOptions.c
===================================================================
--- edk2-0~20180205.c0d9813c.orig/MdeModulePkg/Universal/DisplayEngineDxe/ProcessOptions.c
+++ edk2-0~20180205.c0d9813c/MdeModulePkg/Universal/DisplayEngineDxe/ProcessOptions.c
@@ -829,6 +829,7 @@ PasswordProcess (
     //
     Status = ReadString (MenuOption, gPromptForPassword, StringPtr);
     if (EFI_ERROR (Status)) {
+      ZeroMem (StringPtr, (Maximum + 1) * sizeof (CHAR16));
       FreePool (StringPtr);
       return Status;
     }
@@ -846,7 +847,7 @@ PasswordProcess (
       } else {
         Status = EFI_SUCCESS;
       }
-
+      ZeroMem (StringPtr, (Maximum + 1) * sizeof (CHAR16));
       FreePool (StringPtr);
       return Status;
     }
@@ -862,6 +863,7 @@ PasswordProcess (
     // Reset state machine for password
     //
     Question->PasswordCheck (gFormData, Question, NULL);
+    ZeroMem (StringPtr, (Maximum + 1) * sizeof (CHAR16));
     FreePool (StringPtr);
     return Status;
   }
@@ -877,6 +879,8 @@ PasswordProcess (
     // Reset state machine for password
     //
     Question->PasswordCheck (gFormData, Question, NULL);
+    ZeroMem (StringPtr, (Maximum + 1) * sizeof (CHAR16));
+    ZeroMem (TempString, (Maximum + 1) * sizeof (CHAR16));
     FreePool (StringPtr);
     FreePool (TempString);
     return Status;
